cli/file_watcher.rs

```diff
@@ -61,10 +91,32 @@ where
   }
 }
 
-pub async fn file_watcher(paths: &[PathBuf]) -> Result<(), AnyError> {
-  let (sender, mut receiver) =
-    mpsc::channel::<Result<NotifyEvent, AnyError>>(16);
-  let sender = std::sync::Mutex::new(sender);
+async fn wait_for_file_change(
+  receiver: &Mutex<Receiver<Result<NotifyEvent, AnyError>>>,
+  debounce: &mut Debounce,
+) -> Result<(), AnyError> {
+  while let Some(result) = receiver.lock().unwrap().next().await {
```

Comment:
Good start @bartOssh, but the problem is that if another fs event isn't received the `debounce` won't return anything and the program will hang on this line. I think `Debounce` should have some kind of channel that you can await on. Then in the `Debounce` implementation you should add a tokio timeout that will send a message over that channel.

I think these resources should be helpful:
- https://stackoverflow.com/questions/58651369/how-can-i-split-debounce-and-join-streams
- https://docs.rs/tokio/0.2.22/tokio/time/fn.throttle.html (note this is throttle, which will still return all items from the stream, and we want to return last event in the "debouce" period)

Reply:
Something along the lines (pseudo code):
```rust
struct Debounce {
   ...
}

impl Stream for Debounce {
  fn poll(&self, cx) {
      select! {
           self.delay => {
              // if delay has expired return last seen event
              self.last_event
           },
           event = self.fs_events => {
              self.last_event = event;
              if self.delay.is_none() {
                   // setup delay to fire in 200ms
              } else {
                  // wait for already set up delay to fire
              }
           }
      }
  }
}
```

Reply:
THX a lot. Will try to solve this accordingly.

Reply:
1. I couldn't `await` or `select!` in `poll_next` so I have used looping with `try_recv`. I think provided above example refers to older version of `futures::stream::Stream`  
2. I have removed `select!` as this has been causing race condition. After removing it `watch_func` started working in sync. As an effect `wait_for_file_change` is called only once after spawn deno process finishes its job (func closure).

I was trying to follow Your logic @bartlomieju, as close as my knowledge allowed me. Thanks for good example. I have learned a lot. Hope it is closer to the expected outcome.